### DISCUSSION TITLE
Composer.on is deprecated

### DISCUSSION THREAD

**Discussion by AndrewAntoniuk** [telegraf/telegraf#1741] [Q&A] [answered]
`Composer::on` is deprecated, and it suggests using `Composer::filter`, which is also deprecated, and `filter` suggests using `Composer.drop`.
How can we narrow down the types by ourselves? 
(Eg, I want to have a photo from the message, so I need to be sure that this message contains a photo)
Is this a good way do to so?
``` typescript
const isWithPhoto = (msg: Message & { photo?: PhotoSize[]; }): msg is Message.PhotoMessage => {
    return !!msg.photo;
};
```
I believe telegraf already has some of this logic under the hood :thinking:

**Comment by MKRhere** [comment, marked as answer]
@AndrewAntoniuk It suggests using `filters`, different from `Composer::filter`. Apologies for the confusing wording, and I'll try to link it to some docs in the next release. Read more about why this change was necessary, and how to use filters in the 4.11.0 [release notes](https://github.com/telegraf/telegraf/releases/tag/v4.11.0).

`Composer::filter` was deprecated for a different reason. It doesn't do what you think it would, and is just the inverse of `Composer::drop` (which in fact drops updates that match the condition).

**Reply by AndrewAntoniuk** [reply]
So the actual `bot.on` isn't deprecated if in use with any filter from `telegraf/filters`?

**Reply by MKRhere** [reply]
Correct. Only using `bot.on` and `Composer::on` with message types (`"text"`) are deprecated. Update types (`"message"`, `"callback_query"`) and filters (`message("text")`) are supported.

**Comment by impuLssse** [comment]


**Reply by impuLssse** [reply]
Note: ```All scenes work fine, except for the `on` and `hears` methods```

**Comment by vladkras** [comment]
for anyone facing the same warning with payments, the replacement with filters is
```
bot.on(message('successful_payment'), callback_func)
```
